Resolve issue #25

mtrberzi opened the pull request:
So here are the new Variable and Scope classes, along with some friends (such as Expression and Value) that will be implemented later.

A few things still need to be fixed up (e.g. namespaces) but until we have a better idea of how those will work and behave with our ASTs, I'd rather just check this in and worry about it once that becomes necessary.

Similarly, there are no test cases for this branch because much of the infrastructure we need to write them (e.g. Expressions that can be eval'd and return a type) doesn't yet exist.

mtrberzi committed:
initial AST entities

mtrberzi committed:
tuple expressions

mtrberzi committed:
Merge branch 'master' of file:///home/mtrberzi/whdl/public

mtrberzi committed:
add Variable and Scope classes
todo: exceptions, test cases

mtrberzi committed:
add exceptions for Variable, Scope, and friends

mtrberzi committed:
delete old frontend code, for cleanup purposes

mtrberzi committed:
Merge branch 'master' of file:///home/mtrberzi/whdl/public

Conflicts:
	src/main/java/org/whdl/frontend/syntaxtree/TypeValue.java
	src/main/java/org/whdl/frontend/syntaxtree/Value.java

TysonAndre commented on src/main/java/org/whdl/frontend/syntaxtree/Variable.java:
```diff
@@ -0,0 +1,40 @@
+		return (TypeValue)val;
+	}
+	
+	private boolean assigned = false;
```
nit: put these at the top with the rest of the instance variables

TysonAndre commented:
oops, I was looking at just the first commit when reviewing.

TysonAndre commented on src/main/java/org/whdl/frontend/syntaxtree/VariableIdentifier.java:
```diff
@@ -14,8 +14,19 @@ public VariableIdentifier(NamespaceIdentifier namespaceIdentifier,
   public NamespaceIdentifier getNamespaceIdentifier() {
     return namespaceIdentifier;
   }
+  
```
probably needs hashCode and equals, or a note explaining why it doesn't.
NamespaceIdentifier will as well.

TysonAndre commented on src/main/java/org/whdl/frontend/syntaxtree/CompilationUnit.java:
```diff
@@ -0,0 +1,24 @@
+	public List<Statement> getStatements(){return statements;}
+	
+	public void populateScope(Scope s){
+		
```
``` java
// FIXME implement this
```

lucaswoj replied:
Why wouldn't the `CompilationUnit` have a scope as a private attribute?

lucaswoj commented:
You can merge this pull request with an issue by adding a "#25" in one of the commit messages. Feel free to `push --force` to make that fix :-)

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/CompilationUnit.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.List;
+
+public class CompilationUnit {
```
Why are we calling this `CompilationUnit` over `File` or something? I find the name "CompilationUnit" somewhat vague and confusing.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/CompilationUnit.java:
```diff
@@ -0,0 +1,24 @@
+public class CompilationUnit {
+	private NamespaceIdentifier namespace;
+	private List<NamespaceIdentifier> imports;
+	private List<Statement> statements;
```
Could this perhaps be a `Set` over a `List`? We want to try to stay away from depending on the order of statements.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Expression.java:
```diff
@@ -0,0 +1,6 @@
+package org.whdl.frontend.syntaxtree;
+
+public abstract class Expression {
+	public abstract Value eval();
```
nitey nit nit: Can we just write out the word `evaluate`?

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/MultipleAssignmentException.java:
```diff
@@ -0,0 +1,14 @@
+public class MultipleAssignmentException extends Exception {
+	private static final long serialVersionUID = 1L;
+
+	private Variable var;
```
nit: Just write out the word `variable`.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/MultipleDefinitionException.java:
```diff
@@ -0,0 +1,15 @@
+public class MultipleDefinitionException extends Exception {
+	private static final long serialVersionUID = 1L;
+	
+	private VariableIdentifier id;
```
nit: Just write out the word `identifier`

EDIT: I'm fine with local variables being abbreviated or one letter but class-level things should err on the side of being more explicit.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/CompilationUnit.java:
```diff
@@ -0,0 +1,24 @@
+	public List<NamespaceIdentifier> getImports(){return imports;}
+	public List<Statement> getStatements(){return statements;}
+	
+	public void populateScope(Scope s){
```
nit: should we have a space between the `)` and `{`? We should find an open set of coding standards. I don't care either way -- I'd just like some consistency.

TysonAndre replied:
seconded (also after method definitions) _and control structures like if/else_,

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,52 @@
+	
+	private Map<VariableIdentifier, Variable> symbolTable = new HashMap<VariableIdentifier, Variable>();
+	
+	public void defineVariable(VariableIdentifier identifier, Expression typeExpression) throws MultipleDefinitionException{
```
This line looks too long. Let me find some coding standards that solidify a length requirement.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,52 @@
+
+public class Scope {
+	private Scope parentScope;
+	public Scope(Scope parentScope){
```
nit: put an empty newline between definitions

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,52 @@
+	}
+	public Scope getParentScope(){return parentScope;}
+	
+	private Map<VariableIdentifier, Variable> symbolTable = new HashMap<VariableIdentifier, Variable>();
```
What Tyson said below https://github.com/whdl/whdl/pull/38/files#r7708284

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,52 @@
+		// FIXME this does not handle namespaces correctly
+		if(symbolTable.containsKey(identifier)){
+			// FIXME this does not handle functions with different type signatures correctly
+			throw new MultipleDefinitionException(identifier);
```
One way we could handle this is by having an `UnspecifiedFunctionTypeValue` class that can hold a function with multiple type signatures. 

EDIT: On further consideration, maybe this isn't a great approach for a hardware synthesis language.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,52 @@
+		symbolTable.put(identifier, v);
+	}
+	
+	public TypeValue getVariableTypeValue(VariableIdentifier identifier) throws TypeMismatchException, VariableNotDefinedException{
```
I guess we're just calling these sorts of methods `getVariableType` now.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,52 @@
+			if(parentScope == null){
+				throw new VariableNotDefinedException(identifier);
+			}
+			return parentScope.getVariable(identifier);
```
Throw this in an `else` clause

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Variable.java:
```diff
@@ -0,0 +1,40 @@
+	public VariableIdentifier getIdentifier(){return identifier;}
+	public TypeValue getTypeValue() throws TypeMismatchException{
+		// FIXME if it is possible to type-check this expression earlier, it should be done as soon as possible
+		// e.g. in a semantic analysis pass
```
Cool. We could add a `verify` method to this class which will get called during semantic analysis.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/VariableNotAssignedException.java:
```diff
@@ -0,0 +1,15 @@
+public class VariableNotAssignedException extends Exception {
+	private static final long serialVersionUID = 1L;
+	
+	private Variable var;
```
`var` -> `variable`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/VariableNotDefinedException.java:
```diff
@@ -0,0 +1,14 @@
+public class VariableNotDefinedException extends Exception {
+	private static final long serialVersionUID = 1L;
+
+	private VariableIdentifier id;
```
`id` -> `identifier`

lucaswoj commented:
Despite all my naggy comments, I think this looks really great! It shouldn't be too much work to get this ready to merge.

lucaswoj commented:
... oh, and test cases :smile:

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/VariableIdentifier.java:
```diff
@@ -14,8 +14,19 @@ public VariableIdentifier(NamespaceIdentifier namespaceIdentifier,
+		  return getName();
+	  }else{
+		  // FIXME this is syntax-dependent
+		  return nsString + "." + getName();
```
Wouldn't it still be a `:` between the namespace identifier and variable? For example, `cout` is a member of `std` and we write `std::cout`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/VariableIdentifier.java:
```diff
@@ -14,8 +14,19 @@ public VariableIdentifier(NamespaceIdentifier namespaceIdentifier,
+	  String nsString = getNamespaceIdentifier().toString();
+	  if(nsString.isEmpty()){
+		  return getName();
+	  }else{
```
`} else {`

TysonAndre commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,52 @@
+			// FIXME this does not handle functions with different type signatures correctly
+			throw new MultipleDefinitionException(identifier);
+		}
+		Variable v = new Variable(identifier, typeExpression);
```
add a comment on whether or not variable shadowing is allowed
e.g. this example

```
Boolean b;
if (true) {
    Tuple<Boolean, Integer> b;
}
```

mtrberzi committed:
add hashCode() and equals() to NamespaceIdentifier

mtrberzi committed:
move members to top of class in Variable

mtrberzi committed:
add file scope to CompilationUnit
by the way, this entire branch deals with #25

mtrberzi committed:
change List to Set in CompilationUnit

mtrberzi committed:
change eval() to evaluate()

mtrberzi committed:
fix multiple nits on #25

mtrberzi commented:
Right, thanks for the code review -- new commits are now available. Look again and pull if satisfied.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/NamespaceIdentifier.java:
```diff
@@ -13,4 +13,46 @@ public NamespaceIdentifier(List<String> name) {
+  }
+  
+  @Override
+  public int hashCode(){
```
Is there a library we can call on to generate this hash code? This is one area where my Java knowledge is lacking.

mtrberzi replied:
Not as far as I know -- it needs to be a function of the relevant state, and usually this is done by hand.

lucaswoj replied:
I know that you need to explicitly specify the input to the hash function by hand but are there libraries to generate the hash from that input? Or could you just use the `hasCode` of the list?

lucaswoj replied:
Upon further research, I agree that this is consistent with best practices. Looks good.

mtrberzi committed:
fix nonsensical Scope parameter on CompilationUnit.populateScope()

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/NamespaceIdentifier.java:
```diff
@@ -13,4 +13,46 @@ public NamespaceIdentifier(List<String> name) {
+  }
+  
+  @Override
+  public boolean equals(Object aThat){
```
Why `aThat` over `that`? Just curious

mtrberzi replied:
Eh, it's just a thing I saw, because I use `that` lower down. Usually you give the longer name to the base object and then after doing instanceof checks you can very cutely compare "this and that".

lucaswoj replied:
Cool. Sounds good to me!

lucaswoj commented:
Looks great @MurphEngineer :+1: This just needs some test cases and I'll be ready to merge. 

As a point of general feedback, I think you could be more liberal in your use of vertical whitespace to group related chunks of code within methods. And @TysonAndre and I would prefer if you always had a space before the opening curly brace `{` in a block and before the opening parenthesis in `if`, `while`, etc statements (`if(` -> `if (`)

mtrberzi commented:
(Cool, Ctrl-Shift-F performs exactly the reformatting you like...will have to remember that one.)

Test cases on the way in a bit; I have to bluff to some extent and create facade classes to hide stuff we don't have or don't want to pull into our tests, e.g. complicated Expression stuff

mtrberzi committed:
reformatting of Variable, no syntactic changes

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/NamespaceIdentifier.java:
```diff
@@ -13,4 +13,46 @@ public NamespaceIdentifier(List<String> name) {
+		  if(!nThis.equals(nThat)) return false;
+	  }
+	  return true;
+  }
```
Couldn't you just do `return this.getName().equals(that.getName())` and leverage the built-in `List::equals` method?

mtrberzi replied:
oh yeaaaaah.

mtrberzi committed:
fix suboptimal list equality check in equals()

lucaswoj commented:
Sounds good! Looking forward to it, Murphy. The first code review is always the most painful...

mtrberzi committed:
add equals() to VariableIdentifier

lucaswoj commented:
Don't worry too much about building some testing facades that we're just going to throw away once we get real `Expression` classes. That's a legitimate reason to skip some tests in the short run! Alternately, you could implement a simple `LiteralExpression` class.

mtrberzi committed:
test cases for Variable

mtrberzi committed:
add test for VariableIdentifier.equals()

mtrberzi committed:
reorder definitions

mtrberzi committed:
add hashCode() to VariableIdentifier
This is absolutely necessary as Scope stores V-IDs in a hash table
and we get bogus hashes any other way

mtrberzi committed:
add test cases for Scope

mtrberzi commented:
New test cases are now available.

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/CompilationUnit.java:
```diff
@@ -0,0 +1,27 @@
+package org.whdl.frontend.syntaxtree;
```
In the future, it'll make pull requests easier if you just tackle one issue per pull request.

mtrberzi replied:
I can't actually remember why I did this...this is clearly not being used from anywhere else in the stuff I actually set out to write

lucaswoj replied:
No problem :smile: My brain is mush too

lucaswoj replied:
You might choose to drop this file from this pull request and deal with tests later

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/NamespaceIdentifier.java:
```diff
@@ -13,4 +13,40 @@ public NamespaceIdentifier(List<String> name) {
+	  if(!(aThat instanceof NamespaceIdentifier)) return false;
+	  NamespaceIdentifier that = (NamespaceIdentifier)aThat;
+	  List<String> thisName = this.getName();
+	  List<String> thatName = that.getName();
```
These two lines are dead code now

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Scope.java:
```diff
@@ -0,0 +1,70 @@
+		return parentScope;
+	}
+
+	private Map<VariableIdentifier, Variable> symbolTable = new HashMap<VariableIdentifier, Variable>();
```
Please move all property declarations to the top of the class

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/Variable.java:
```diff
@@ -0,0 +1,50 @@
+	private Expression typeExpression;
+
+	private boolean assigned = false;
+	private Expression valExpr;
```
`valExpr` -> `valueExpression`

lucaswoj commented on src/main/java/org/whdl/frontend/syntaxtree/VariableIdentifier.java:
```diff
@@ -2,20 +2,48 @@
+	
+	@Override
+	public boolean equals(Object aThat) {
+		if (this == aThat)
```
Please either use curly braces `{}` or put the `return true` on the same line as the `if (this == aThat)`.

lucaswoj commented on src/test/java/org/whdl/frontend/syntaxtree/TestScope.java:
```diff
@@ -0,0 +1,138 @@
+	}
+
+	@Test
+	public void testAssignVariable() throws MultipleDefinitionException,
```
Its best to just do one atomic "test" per method in this class. I'd probably break this up into 
- `testAssignVariable`
- `testAssignVariableMultipleAssignment`
- `testAssignVariableNotDefined`

Also note that junit has some built-in exception checkers http://stackoverflow.com/questions/156503/how-do-you-assert-that-a-certain-exception-is-thrown-in-junit-4-tests

lucaswoj commented on src/test/java/org/whdl/frontend/syntaxtree/TestVariable.java:
```diff
@@ -0,0 +1,109 @@
+		}
+		// and guess what we can ALSO use as the value expression...
+		try{
+			v.setValue(getTypeExpression());
```
This is risky... It might suddenly start failing later when we add verification

lucaswoj commented on src/test/java/org/whdl/frontend/syntaxtree/TestVariable.java:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	@Test
+	public void testGetValue() {
```
Again, I'd break this up into one method per assertion
- `testGetValue`
- `testGetValueNotInitiallyAssigned`
- `testGetValueNotAssignedException`
- `testGetValueNotAssignedException`

lucaswoj commented on src/test/java/org/whdl/frontend/syntaxtree/TestVariable.java:
```diff
@@ -0,0 +1,109 @@
+		// implicitly tests Variable.isAssigned(), Variable.setValue()
+		Variable v = getVariable();
+		assertFalse("newly created variable incorrectly appears to be assigned", v.isAssigned());
+		try{
```
should have spaces around curly braces here (Peter Bhur style (but only for curly braces))

lucaswoj commented on src/test/java/org/whdl/frontend/syntaxtree/TestVariable.java:
```diff
@@ -0,0 +1,109 @@
+	}
+
+	@Test
+	public void testSetValue() {
```
Break this one up

lucaswoj commented on src/test/java/org/whdl/frontend/syntaxtree/TestVariableIdentifier.java:
```diff
@@ -34,4 +34,23 @@ public void testGetNamespaceIdentifier() {
   }
+  
+  @Test
+  public void testEquals(){
```
This one too

lucaswoj commented:
Looks like you're using tabs here. I've started off the codebase using two spaces.

lucaswoj commented:
I feel bad that this has turned into such an ordeal (this is why I keep reiterating that every pull request should be as small and focused as possible) so I fixed all of the outstanding comments on this pull request https://gist.github.com/lucaswoj/e251b4a459a28b8bad4c

BUT there are still a bunch of lines lacking test coverage. If you run 

```
gradle jacocoTestReport && open ./build/reports/jacoco/jacocoTestReport/index.html
```

you can see what's missing

mtrberzi committed:
address multiple nits

mtrberzi commented:
Don't feel bad, I'd rather spend time on this and get it right than charge ahead and not learn anything about how it should be done.

Will fix up the test cases and code coverage in a bit.

mtrberzi committed:
abstract out namespace separator and add test case

mtrberzi committed:
add test case for parent scope name resolution
we now have 100% coverage of Scope in this pull request

mtrberzi commented:
New test cases are available...for Scope. The missed branch in Variable is due to not having the infrastructure to perform a type check elsewhere, and I'm not really planning on creating test infrastructure for an exceptional condition that will probably be moved elsewhere later on.

mtrberzi committed:
multiple fixes from lucas
Most of the noise here is due to renaming Variable.getTypeValue()
to Variable.getType()
Also test case improvements, thanks much

mtrberzi commented:
New test cases are available.

mtrberzi committed:
delete CompilationUnit, it is not in the scope of this pull request
